File: PhotoViewer-ObjC/SessionDelegate.m
//
//  SessionDelegate.m
//  PhotoViewer
//
//  Created by ASIM27 on 1/20/17.
//  Copyright © 2017 km3h. All rights reserved.
//

#import "SessionDelegate.h"

@interface SessionDelegate ()
@property (strong, nonatomic) NSMutableData *dataReceived;
@property (nonatomic, copy, nullable) void (^completionHandler) (NSData*, NSURLResponse*, NSError*);
@end

@implementation SessionDelegate

-(instancetype)initWithCompletionHandler:(void (^) (NSData*, NSURLResponse*, NSError*)) completionHandler
{
    self = [super init];
    self.completionHandler = completionHandler;
    self.dataReceived = [[NSMutableData alloc] init];
    return self;
}

-(void)URLSession:(NSURLSession *)session dataTask:(NSURLSessionDataTask *)dataTask didReceiveData:(NSData *)data
{
    [self.dataReceived appendData:data];
}

-(void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didCompleteWithError:(NSError *)error
{
    self.completionHandler(self.dataReceived, task.response, nil);
}

@end

File: PhotoViewer/SessionDelegate.swift
//
//  SessionDelegate.swift
//  PhotoViewer
//
//  Created by ASIM27 on 11/30/16.
//  Copyright © 2016 km3h. All rights reserved.
//

import UIKit

class SessionDelegate: NSObject
{
    var dataReceived = Data()
    typealias handlerName = (Data?, URLResponse?, Error?) -> Void
    var handler : handlerName?
    
    required init?(handler : handlerName?)
    {
        super.init()
        self.handler = handler
    }
}

extension SessionDelegate : URLSessionDelegate
{
    func urlSession(_ session: URLSession, didBecomeInvalidWithError error: Error?)
    {
    }
    
    func urlSession(_ session: URLSession, didReceive challenge: URLAuthenticationChallenge, completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void)
    {
        completionHandler(.performDefaultHandling, nil)
    }
    
    func urlSessionDidFinishEvents(forBackgroundURLSession session: URLSession)
    {
        
    }
}

extension SessionDelegate : URLSessionTaskDelegate
{
    func urlSession(_ session: URLSession, task: URLSessionTask, willPerformHTTPRedirection response: HTTPURLResponse, newRequest request: URLRequest, completionHandler: @escaping (URLRequest?) -> Void)
    {
        completionHandler(request)
    }
    
    func urlSession(_ session: URLSession, task: URLSessionTask, didReceive challenge: URLAuthenticationChallenge, completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void)
    {
        completionHandler(.performDefaultHandling, nil)
    }
    
    func urlSession(_ session: URLSession, task: URLSessionTask, needNewBodyStream completionHandler: @escaping (InputStream?) -> Void)
    {
    }
    
    func urlSession(_ session: URLSession, task: URLSessionTask, didSendBodyData bytesSent: Int64, totalBytesSent: Int64, totalBytesExpectedToSend: Int64)
    {
    }
    
    func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?)
    {
        if let handler = self.handler
        {
            handler(self.dataReceived, task.response, nil)
        }
    }
    
    func urlSession(_ session: URLSession, task: URLSessionTask, didFinishCollecting metrics: URLSessionTaskMetrics)
    {
        
    }
}

extension SessionDelegate: URLSessionDataDelegate
{
    func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didReceive response: URLResponse, completionHandler: @escaping (URLSession.ResponseDisposition) -> Void)
    {
        completionHandler(.allow)
    }
    
    func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didBecome downloadTask: URLSessionDownloadTask)
    {
    }
    
    func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didBecome streamTask: URLSessionStreamTask)
    {
    }
    
    func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didReceive data: Data)
    {
        dataReceived.append(data)
        print(dataReceived.count)
    }
    
    func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, willCacheResponse proposedResponse: CachedURLResponse, completionHandler: @escaping (CachedURLResponse?) -> Void)
    {
        completionHandler(proposedResponse)
    }
}
